### PR TITLE
DataAccess Refactoring

### PR DESCRIPTION
**Pull request by hishamco**

### CODE REVIEW

File: src/OrchardCore/Orchard.Data.Abstractions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -1,27 +1,24 @@
...
 
 namespace Orchard.Data
 {
     public static class ServiceCollectionExtensions
     {
-        public static IServiceCollection TryAddDataProvider(this IServiceCollection services, string name, string value, bool hasConnectionString)
+        public static IServiceCollection TryAddDataProvider(this IServiceCollection services, string name, string value, bool hasConnectionString = false)
         {
-            for (var i = services.Count - 1; i >= 0; i--)
+            var serviceProvider = services.BuildServiceProvider();
```
Comment (sebastienros) [inline src/OrchardCore/Orchard.Data.Abstractions/ServiceCollectionExtensions.cs]: No way ;) No refactoring can justify the need to build the container early like this.

Comment (hishamco) [inline src/OrchardCore/Orchard.Data.Abstractions/ServiceCollectionExtensions.cs]: I added this at end 😄 I was worrying if may has a side effect, let me revert this change

**Comment by sebastienros** [comment]
I really appreciate the effort you are putting into it but this code is going nowhere.
If there is a clear goal or advantage here please let me know. It's harder to understand and it's hacky, yet can't see any advantage.

Hisham, if you are looking for things to do to help, we have a HUGE backlog of bugs and features that are way more interesting and useful.

**Comment by hishamco** [comment]
I don't know why it's harder 😄, IMHO the LINQ query simplify the things, if you don't like it, I will revert to the original `for` statement plus use pattern matching

> Hisham, if you are looking for things to do to help, we have a HUGE backlog of bugs and features that are way more interesting and useful.

Sounds good 👍

**Comment by Jetski5822** [comment]
@sebastienros what about abstracting this stuff out to like a 

OrchardCore.Data
 -> OrchardCore.Data.SqlLite
 -> OrchardCore.Data.Oracle

that way the providers are separated off and cleaned up. then if we want to support a new one, the other ones arent affected.

**Comment by hishamco** [comment]
That's what I mean @Jetski5822 in the issue #790

**Comment by sebastienros** [comment]
Closing to cleanup PRs, let's open an issue to talk about improvements later.

**Comment by hishamco** [comment]
Ok
